pkg/apis/core/v1alpha1/condition.go

```diff
@@ -17,181 +17,229 @@ limitations under the License.
 package v1alpha1
 
 import (
+	"sort"
+
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// ConditionType type for possible conditions the resource could be in.
+// A ConditionType represents a condition a resource could be in.
 type ConditionType string
 
-// Resource conditions.
+// Condition types.
 const (
-	// Pending means that the resource create request has been received and is waiting to be fulfilled.
-	Pending ConditionType = "Pending"
-	// Creating means that the resource create request has been accepted and the resource is in
-	// the process of being created.
-	Creating ConditionType = "Creating"
-	// Deleting means that the resource is in the process of being deleted.
-	Deleting ConditionType = "Deleting"
-	// Failed means that the resource is in a failure state, for example it failed to be created.
-	Failed ConditionType = "Failed"
-	// Ready means that the resource creation has been successful and the resource is ready to
-	// accept requests and perform operations.
+	// Ready managed resources are believed to be ready to handle work.
 	Ready ConditionType = "Ready"
+
+	// Synced managed resources are believed to be in sync with the Kubernetes
+	// resources that manage their lifecycle.
+	Synced ConditionType = "Synced"
+)
+
+// A ConditionReason represents the reason a resource is in a condition.
+type ConditionReason string
+
+// Condition reasons.
+const (
+	// Available indicates the managed resource is available for use.
+	Available ConditionReason = "Managed resource is available"
+
+	// Unavailable indicates the managed resource is not available for use.
+	Unavailable ConditionReason = "Managed resource is unavailable"
```

Comment:
What is the difference between `Unavailable.Status = true` and `Available.Status = false`?

Reply:
It's not as obvious as it should be, but`ConditionStatus` corresponds to `ConditionType`, not `ConditionReason`. The hierarchy is something like:

1. `ConditionType` - What type of condition is this?
2. `ConditionStatus` - Is the condition true, false, or unknown?
3. `ConditionReason` - Why is the condition true, false or unknown?

So in this PR the subset of combinations we use are:

* `Type=Ready`
  * `Status=True`
    * `Reason=Available`
  * `Status=False`
    * `Reason=Unavailable` - Resource is unready because it's unavailable.
    * `Reason=Creating` - Resource is unready because it's 
    * `Reason=Deleting`
* `Type=Synced`
  * `Status=True`
    * There's actually no corresponding `Reason` here. Perhaps there should be something like `ReconcileSuccess`.
  * `Status=False`
    * `Reason=Error`- Resource was not synced because we encountered an error.

Does that make sense?

Reply:
I just pushed a commit that tries to make this a little more obvious by adding docstrings to the fields of the `Condition` struct.

Reply:
I'm going to mark this resolved - feel free to reopen if you have more thoughts!